### PR TITLE
Added Aversive++ package.

### PR DESCRIPTION
**Pull request by astralien3000**
Hello,

It is my first PR here, so I don't know if what I am doing is ok. I added an external library called [Aversive++](https://github.com/AversivePlusPlus/AversivePlusPlus), which is a library for developping on microcontrollers, but specialized in robotics.

The (tiny) part of the library that is currently available to RIOT is the core module (basic types), the containers (lists, buffers, fifo, lifo), the stream interface (similare to STL streams), and a driver for some servomotors (SCS15 from Feetech).

I had a technical problem : the RIOT build system needs an "aversiveplusplus.a" archive to be created, but since the part of the library that is used is header-only, there is no need of compiling anything. So that I created an empty one to make the build system happy.

Cheers !

### CODE REVIEW

**Comment by kaspar030** [comment]
Nice!

> I had a technical problem : the RIOT build system needs an "aversiveplusplus.a" archive to be created, but since the part of the library that is used is header-only, there is no need of compiling anything.

Could you try adding "aversiveplusplus"as pseudo module (to Makefile.pseudomodules in the tree root)?

File: pkg/aversiveplusplus/Makefile [hunk 1]
```diff
@@ -0,0 +1,11 @@
+PKG_NAME=aversiveplusplus
+PKG_URL=https://github.com/AversivePlusPlus/AversivePlusPlus
+PKG_VERSION ?= 86e53477ff98a9e111c7cf612f8d248320dadbd7
```
Comment (kaspar030) [inline pkg/aversiveplusplus/Makefile]: As PKG_VERSION is shared between multiple packages, maybe it is safer to keep this as non-conditional assignment. What do you think?

**Comment by astralien3000** [comment]
> Could you try adding "aversiveplusplus"as pseudo module (to Makefile.pseudomodules in the tree root)?

I have already tried this. It did not change the error I had without ("aversiveplusplus.a" not found). Someone told me that I could try also to add "USEMODULE += aversiveplusplus" after "USEPKG += aversiveplusplus" in my makefile, but it did not work too.

And when trying to understand the build system, It seemed to me that the "PSEUDOMODULES" variable only affect modules, and not packages. But maybe I'm wrong since I'm quite new here.

File: pkg/aversiveplusplus/Makefile (hunk 1 above)
Comment (astralien3000) [inline pkg/aversiveplusplus/Makefile]: Ok, sure, I don't even know why I wrote it this way. Maybe a bad copy/past ^^

**Comment by kaspar030** [comment]
> And when trying to understand the build system, It seemed to me that the
> "PSEUDOMODULES" variable only affect modules, and not packages. But
> maybe I'm wrong since I'm quite new here.

You're totally correct. I fixed that in #5671.

**Comment by kaspar030** [comment]
#5671 has been merged, so you can remove the fake .a.

**Comment by astralien3000** [comment]
Ok, thanks ! How do I "merge" the modification ? I just merge the master in my branch or is there a better solution ?

**Comment by miri64** [comment]
```bash
git fetch origin  # given origin is the RIOT-OS remote, and not your own
git rebase -i origin/upstream
```

And delete every line in the editor that isn't your own commit. Then we still are able to track your commits.

**Comment by miri64** [comment]
oh and you have to force-push to your repo then:

```bash
git push -f astralien3000 dev-aversiveplusplus # given astralien3000 is the remote of your repo
```

**Comment by aabadie** [comment]
In this case, I don't think the `-i` option is needed.
BTW, you can do fetch and rebase in a single line:
```
git pull --rebase origin/master
```

**Comment by astralien3000** [comment]
Thank you ! I have done it the first way, but I'll try the second one next time ! ^^

**Comment by astralien3000** [comment]
Hey !
Before including it in my pull request, I wanted to show you the [first Aversive++ driver that links to a RIOT driver](https://github.com/astralien3000/stream-riot/blob/master/include/stream/riot/uart_stream.hpp).

It is juste a class that convert the [periph/uart](https://github.com/RIOT-OS/RIOT/blob/master/drivers/include/periph/uart.h) interface to an [Aversive++ Stream](https://github.com/astralien3000/test_riot_buggybot/blob/master/buggybot_server/main.cpp), in order to be used with the SCS15 driver. There is an example of use of RIOT and Aversive++ together [here](https://github.com/astralien3000/test_riot_buggybot/blob/master/buggybot_server/main.cpp).

If you have any feedback about this ^^

**Comment by astralien3000** [comment]
So, the code I wanted to propose (uart stream) has been tested on a samr21-xpro board with my robot, and works, so that I add it.

File: pkg/aversiveplusplus/Makefile.include [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+INCLUDES += -I$(BINDIRBASE)/pkg/$(BOARD)/aversiveplusplus/modules/device-control/include
+INCLUDES += -I$(BINDIRBASE)/pkg/$(BOARD)/aversiveplusplus/modules/device-2wheel/include
+
+INCLUDES += -I$(BINDIRBASE)/pkg/$(BOARD)/aversiveplusplus/modules/cas/include
+INCLUDES += -I$(BINDIRBASE)/pkg/$(BOARD)/aversiveplusplus/modules/ik/include
+
+INCLUDES += -I$(BINDIRBASE)/pkg/$(BOARD)/aversiveplusplus/modules/feetech/include
+
+# The package is not compiled (header-only)
+PSEUDOMODULES += aversiveplusplus
```
Comment (PeterKietzmann) [inline pkg/aversiveplusplus/Makefile.include]: Please add a newline here

**Comment by PeterKietzmann** [comment]
@astralien3000 thanks for your efforts and introducing your great project! We'd like to merge your PR soon. Could you squash all commits to one and name it like "Aversive++: initial import"? If you need further help with git or so, just ask here!

**Comment by astralien3000** [comment]
Done !

**Comment by cgundogan** [comment]
@astralien3000 looks like a neat libary!

>Before including it in my pull request, I wanted to show you the first Aversive++ driver that links to a RIOT driver.

I am wondering how you want to integrate the driver into RIOT. Could you draft something up?

**Comment by astralien3000** [comment]
What do you mean ? Do you want that I write some example project and put it in the example directory ?

The driver is already available via this PR, and you can use it [like this](https://github.com/astralien3000/test_riot_buggybot/blob/master/buggybot_server/main.cpp#L41).

**Comment by cgundogan** [comment]
>Do you want that I write some example project and put it in the example directory ?

Not an example application per se, but a test or a unittest application in `RIOT/tests/`, `RIOT/tests/unitests` respectively, would be very nice. This way the pkg will be built by our ci server (murdock).

**Comment by astralien3000** [comment]
Done ! I hope this is ok.

**Comment by cgundogan** [comment]
what is the test supposed to do and how can I verify it's working successfully? (:

**Comment by astralien3000** [comment]
Well, it is a driver for [SCS15 servomotors](http://www.feetechrc.com/product/digital-servo/feetech-scservo-multi-purpose-uart-bus-robot-servo-for-arduino-diy/). It opens the UART 1 communication (at 1M baudrate), and send position commands to the connected motors. The motors should not move because their torque is disabled by default.
So that, you need the servomotors to test is the code works well. And even with that, the test should be modified to enable the servomotor's torque.

So I guess you would prefer a simpler test with only the UART communication ?

**Comment by cgundogan** [comment]
I sniffed the uart output. However, I cannot say anything about the protocol being used. I see bytes over the wire, but I cannot judge whether they are correct. I guess I have to rely on the correctness of `Aversive++` here (:

**Comment by astralien3000** [comment]
And the problem with these servomotors it that there is no "official" documentation of the protocol ><, I made my code from a dirty/non commented Arduino library.
But it should be ok, since I have [tested it with my robot](https://www.youtube.com/watch?v=J5Ogdffptdg).

**Comment by emmanuelsearch** [comment]
maybe @kYc0o can have access to the hardware and test if it works?

**Comment by kYc0o** [comment]
Yes I can try.

**Comment by astralien3000** [comment]
Ha ? Murdock does not seem to be happy. I can't tell where I did something wrong.
@kYc0o, you can get the servomotors anytime today.

**Comment by kaspar030** [comment]
```
BUILD OUTPUT of group static-tests:
Current branch HEAD is up to date.
Current branch HEAD is up to date.
tests/pkg_aversiveplusplus/main.cpp:15: trailing whitespace.
+  
ERROR: This change introduces new whitespace errors
file has an unknown license header: 'tests/pkg_aversiveplusplus/main.cpp'
```

File: pkg/aversiveplusplus/Makefile.include [hunk 3]
```diff
@@ -0,0 +1,18 @@
+INCLUDES += -I$(BINDIRBASE)/pkg/$(BOARD)/aversiveplusplus/modules/base/include
```
Comment (kaspar030) [inline pkg/aversiveplusplus/Makefile.include]: would you mind putting ```$(BINDIRBASE)/pkg/$(BOARD)/aversiveplusplus/modules``` into a variable and use that throughout this file?

File: pkg/aversiveplusplus/Makefile.include [hunk 4]
```diff
@@ -0,0 +1,20 @@
+MODULES = $(BINDIRBASE)/pkg/$(BOARD)/aversiveplusplus/modules
```
Comment (miri64) [inline pkg/aversiveplusplus/Makefile.include]: I think it's safer to name that variable `AVERSIVEPLUSPLUS_MODULES` or something like this.

**Comment by kYc0o** [comment]
I just tried this PR with success on samr21-xpro and arduino-mega2560. Some problems were found on arduino-due and waspmote-pro, possibly related to the high baudrate used by the test on a servomotor.

I've asked @astralien3000 to check if at lower baudrates it works, but I ACK this PR as is so it can be merged.

Good job!

**Comment by kYc0o** [comment]
Oh, of course the ACK comes when the previous comments are addressed.

**Comment by kYc0o** [comment]
@miri64 Can you confirm that your requested changes were addressed? So we can merge this ;)

**Comment by kYc0o** [comment]
Yeah! Thanks ;) and GO!

**Pull request merged**